src/ball_sort/tube.cpp

```diff
@@ -0,0 +1,73 @@
+#include "ball_sort/tube.hpp"
+#include <iostream>
+#include <string>
+#include <unordered_set>
+
+Tube::Tube(const std::string &balls) : m_balls{balls} {}
+
+auto Tube::is_empty() const -> bool
+{
+    return m_balls.empty();
+}
+
+auto Tube::is_full() const -> bool
+{
+    return m_balls.size() == MAX_CAPACITY;
+}
+
+auto Tube::is_one_colour() const -> bool
+{
+    std::unordered_set<char> ball_colours{};
+    for (const char &ball : m_balls) {
+        ball_colours.insert(ball);
+    }
```

Comment:
slow, just check if they are all equal to the first one. (It's quicker to set(...) in python because C does that under the hood)